clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskDefinitionCachingAgent.java

```diff
@@ -90,71 +81,55 @@ public String getAgentType() {
 
   @Override
   protected List<TaskDefinition> getItems(AmazonECS ecs, ProviderCache providerCache) {
-    List<TaskDefinition> taskDefinitionList = new LinkedList<>();
-    Set<String> cachedArns =
-        providerCache.getIdentifiers(TASK_DEFINITIONS.toString()).stream()
-            .filter(id -> keyAccountRegionFilter(TASK_DEFINITIONS.toString(), id))
-            .map(
-                id -> {
-                  Map<String, String> keyParts = Keys.parse(id);
-                  return keyParts.get("taskDefinitionArn");
-                })
-            .collect(Collectors.toSet());
-
-    String nextToken = null;
-    do {
-      ListTaskDefinitionsRequest listTasksRequest = new ListTaskDefinitionsRequest();
-      if (nextToken != null) {
-        listTasksRequest.setNextToken(nextToken);
-      }
-      ListTaskDefinitionsResult listTaskDefinitionsResult =
-          ecs.listTaskDefinitions(listTasksRequest);
-      List<String> taskDefinitionArns = listTaskDefinitionsResult.getTaskDefinitionArns();
+    // get all ECS services in region for account
+    ServiceCacheClient serviceCacheClient = new ServiceCacheClient(providerCache, objectMapper);
+    Collection<Service> services = serviceCacheClient.getAll(accountName, region);
+    log.debug("Found {} ECS services for which to cache task definitions", services.size());
 
-      if (taskDefinitionArns.size() == 0) {
-        continue;
-      }
+    Set<String> taskDefArns = new HashSet<>();
 
-      Set<String> newTaskDefArns = new HashSet<>(taskDefinitionArns);
-      newTaskDefArns.removeAll(cachedArns);
+    for (Service service : services) {
+      taskDefArns.add(service.getTaskDefinition());
+    }
 
-      Set<String> existingTaskDefArns = new HashSet<>(taskDefinitionArns);
-      existingTaskDefArns.removeAll(newTaskDefArns);
+    List<TaskDefinition> taskDefinitions = new ArrayList<>();
 
-      if (!existingTaskDefArns.isEmpty()) {
-        // TaskDefinitions are immutable, there's no reason to make a describe call on existing
-        // ones.
-        taskDefinitionList.addAll(retrieveFromCache(existingTaskDefArns, providerCache));
-      }
+    int newTaskDefs = 0;
+
+    for (String arn : taskDefArns) {
 
-      for (String taskDefinitionArn : newTaskDefArns) {
-        TaskDefinition taskDefinition =
-            ecs.describeTaskDefinition(
-                    new DescribeTaskDefinitionRequest().withTaskDefinition(taskDefinitionArn))
+      // TaskDefinitions are immutable, there's no reason to
+      // make a describe call on existing ones.
+      TaskDefinition cacheEntry = retrieveFromCache(arn, providerCache);
+
+      if (cacheEntry != null) {
+        taskDefinitions.add(cacheEntry);
```

Comment:
Is it possible that the cacheEntry is stale?

Reply:
Meaning, is it possible for the cached task def to no longer match the task def in the account w/ the same ARN? 
It is not! Task defs are immutable. Even if you deregister a revision in a family, and then register a new one in the same family, it will be created with a revision number incremented higher than the latest revision, active or not.

Reply:
Good point! So different revisions will be considered different task defs!